File: java/FlowFieldsGen/src/main/java/in/mcxiv/ui/SimpleMutableFloatingColor.java
package in.mcxiv.ui;

import java.awt.*;
import java.awt.color.ColorSpace;

public class SimpleMutableFloatingColor extends Color {

    private static final int REMOVE_TRANSPARENCY_MASK = ~(0xFF << 24);
    private float transparency;

    public SimpleMutableFloatingColor(float r, float g, float b, float a) {
        super(r, g, b, a);
        transparency = a;

        // int_a = (int)(float_a*255+0.5)
        // float_a = (((float)int_a)-0.5f)/255f
    }

    public void setTransparency(float transparency) {
        assert transparency >= 0 && transparency <= 1 : "transparency >= 0 && transparency <= 1 but " + transparency;
        this.transparency = transparency;
    }

    @Override
    public int getAlpha() {
        return (int) (transparency * 255 + 0.5);
    }

    @Override
    public int getRGB() {
        return (super.getRGB() & REMOVE_TRANSPARENCY_MASK) | (getAlpha() << 24);
    }

    @Override
    public float[] getRGBComponents(float[] compArray) {
        float[] rgbComponents = super.getRGBComponents(compArray);
        rgbComponents[3] = transparency;
        return rgbComponents;
    }

    @Override
    public float[] getComponents(float[] compArray) {
        // TODO: do i have to put out transparency again?
        return super.getComponents(compArray);
    }
}

File: static_web/FlowFieldsGen/js/ui/SimpleMutableFloatingColor.js
const REMOVE_TRANSPARENCY_MASK = ~(0xFF << 24);

class SimpleMutableFloatingColor {
  constructor(r, g, b, a) {
    this.r = r;
    this.g = g;
    this.b = b;
    this.a = a;
  }

  intize(float) {
    return Math.floor(float * 255 + 0.5);
  }

  floatize(int) {
    return ((int - 0.5) / 255);
  }

  setTransparency(transparency) {
    console.assert(transparency >= 0 && transparency <= 1, "transparency >= 0 && transparency <= 1 but " + transparency);
    this.a = transparency;
  }

  getRed() {
    return this.intize(this.r);
  }

  getGreen() {
    return this.intize(this.g);
  }

  getBlue() {
    return this.intize(this.b);
  }

  getAlpha() {
    return this.intize(this.a);
  }

  getRGB() {
    // TODO: Is there a faster version?
    return "#" + (this.getRed()).toString(16) + (this.getGreen()).toString(16) + (this.getBlue()).toString(16);
  }

  equals(color) {
    return this.a === color.a;
  }
}
